### PR TITLE
feat(CL Fees): initialize fee growth outside for ticks

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #3956

## What is the purpose of the change

Implementing the initialization for the fee growth outside tee accumulator.

This is extracted from the original fee POC: https://github.com/osmosis-labs/osmosis/pull/3893

## Changelog
- Implement and test `getInitialFeeGrowthOtsideForTick`
- Implement and test  `chargeFee`
- Refactor and test `getTickinfo`

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

### CODE REVIEW

File: x/concentrated-liquidity/tick.go [hunk 1]
```diff
@@ -64,7 +64,14 @@ func (k Keeper) getTickInfo(ctx sdk.Context, poolId uint64, tickIndex int64) (ti
 	found, err := osmoutils.Get(store, key, &tickStruct)
 	// return 0 values if key has not been initialized
 	if !found {
-		return model.TickInfo{LiquidityGross: sdk.ZeroDec(), LiquidityNet: sdk.ZeroDec()}, err
+		// If tick has not yet been initialized, we create a new one and initialize
+		// the fee growth outside,
```
Comment (p0mvn) [inline x/concentrated-liquidity/tick.go:68]:
```suggestion
		// the fee growth outside.
```

File: x/concentrated-liquidity/fees.go [hunk 2]
```diff
@@ -124,6 +143,32 @@ func (k Keeper) getFeeGrowthOutside(ctx sdk.Context, poolId uint64, lowerTick, u
 	return feeGrowthAboveUpperTick.Add(feeGrowthBelowLowerTick...), nil
 }
 
+// getInitialFeeGrowthOtsideForTick returns the initial value of fee growth outside for a given tick.
```
Comment (p0mvn) [inline x/concentrated-liquidity/fees.go:146]:
```suggestion
// getInitialFeeGrowthOutsideForTick returns the initial value of fee growth outside for a given tick.
```

File: x/concentrated-liquidity/fees.go [hunk 3]
```diff
@@ -124,6 +143,32 @@ func (k Keeper) getFeeGrowthOutside(ctx sdk.Context, poolId uint64, lowerTick, u
...
 
+// getInitialFeeGrowthOutsideForTick returns the initial value of fee growth outside for a given tick.
+// This value depends on the tick's location relative to the current tick.
+//
+// feeGrowthOutside = { feeGrowthGlobal current tick >= tick }
+//                    { 0               current tick <  tick }
+//
+// The value is chosen as if all of the fees earned to date had occurrd below the tick.
+// Returns error if the pool with the given id does exist or if fails to get the fee accumulator.
+func (k Keeper) getInitialFeeGrowthOtsideForTick(ctx sdk.Context, poolId uint64, tick int64) (sdk.DecCoins, error) {
```
[the remark this answers is not in the document]
Comment (mattverse) [reply x/concentrated-liquidity/fees.go:154]: and same changes should be applied to all callers of the method

File: x/concentrated-liquidity/tick_test.go [hunk 4]
```diff
@@ -273,17 +277,26 @@ func (s *KeeperTestSuite) TestGetTickInfo() {
...
-			poolToGet:        validPoolId,
-			tickToGet:        DefaultCurrTick.Int64() + 1,
+			name:      "Get tick info on existing pool with no existing tick (cur pool tick > tick)",
+			poolToGet: validPoolId,
+			tickToGet: DefaultCurrTick.Int64() + 1,
+			// Note that FeeGrowthOutside is not initialized.
 			expectedTickInfo: model.TickInfo{LiquidityGross: sdk.ZeroDec(), LiquidityNet: sdk.ZeroDec()},
 		},
+		{
+			name:      "Get tick info on existing pool with no existing tick (cur pool tick == tick)",
```
Comment (mattverse) [inline x/concentrated-liquidity/tick_test.go:294]:
```suggestion
			name:      "Get tick info on existing pool with no existing tick (cur pool tick == tick), initialized fee growth outside",
```

**Review by mattverse — APPROVED** [review]
Straight foward LGTM 🌮 
(Note to change typo before merging)

**Pull request merged**